src/fcntl.rs

```diff
@@ -53,6 +55,23 @@ pub fn openat<P: ?Sized + NixPath>(dirfd: RawFd, path: &P, oflag: OFlag, mode: M
     Errno::result(fd)
 }
 
+pub fn readlink<'a, P: ?Sized + NixPath>(path: &P, buffer: &'a mut [u8]) -> Result<&'a OsStr> {
+    let res = try!(path.with_nix_path(|cstr| {
+        unsafe { libc::readlink(cstr.as_ptr(), buffer.as_mut_ptr() as *mut c_char, buffer.len() as size_t) }
+    }));
+
+    Errno::result(res).map(move |len| OsStr::from_bytes(&buffer[..(len as usize)]))
```

Comment:
ok, this solution is wrong: if the buffer is too small, the buffer is truncated without notifications. I see two options here:

1. Remove buffer from arguments and reallocate buffer internally until it fits the result of readlink (this is what the rust/go stdlib do)
2. return ENAMETOOLONG if len > buffer.len()